core/src/main/scalajvm/scalapb/zio_grpc/server/ZServerCallHandler.scala

```diff
@@ -108,30 +111,62 @@ object ZServerCallHandler {
       call: ZServerCall[Res],
       stream: ZStream[Any, Status, Res]
   ): ZIO[Any, Status, Unit] = {
-    def innerLoop(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Boolean] =
-      queue.take
-        .flatMap {
-          case Exit.Success(res)   => call.sendMessage(res).as(true)
-          case Exit.Failure(cause) =>
-            cause.failureOrCause match {
-              case Left(Some(status)) => ZIO.fail(status)
-              case Left(None)         => ZIO.succeed(false)
-              case Right(cause)       => ZIO.failCause(cause)
+    def takeFromQueue(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Unit] =
+      queue.takeAll.flatMap(takeFromCache(_, queue))
+
+    def takeFromCache(
+        xs: Chunk[Exit[Option[Status], Res]],
+        queue: Dequeue[Exit[Option[Status], Res]]
+    ): ZIO[Any, Status, Unit] =
+      ZIO.suspendSucceed {
+        @tailrec def innerLoop(loop: Boolean, i: Int): IO[Status, Unit] =
+          if (loop) {
+            xs(i) match {
+              case Failure(cause) =>
+                cause.failureOrCause match {
+                  case Left(Some(status)) =>
+                    ZIO.fail(status)
+                  case Left(None)         =>
+                    ZIO.unit
+                  case Right(cause)       =>
+                    ZIO.failCause(cause)
+                }
+              case Success(value) =>
+                call.call.sendMessage(value)
+                // at this point we have consumed i + 1 elements from the chunk
+                val newI = i + 1
+                // the loop iteration may only continue if the call can
+                // still accept elements and we have more elements to send
+                innerLoop(call.call.isReady && newI < xs.length, newI)
             }
-        }
-        .repeatWhileZIO(res => call.isReady.map(_ && res))
+          } else if (call.call.isReady)
+            // ^ if we reached the end of the chunk but the call can still
+            // proceed, we pull from the queue and continue
+            takeFromQueue(queue)
+          else
+            // ^ otherwise, we wait for the call to be ready and then start again
+            outerLoop(Some(xs.drop(i)), queue)
+
+        if (xs.isEmpty)
+          takeFromQueue(queue)
+        else
+          innerLoop(true, 0)
+      }
 
-    def outerLoop(queue: Dequeue[Exit[Option[Status], Res]]): ZIO[Any, Status, Boolean] =
-      (call.awaitReady *> innerLoop(queue))
-        .repeatWhile(identity)
+    def outerLoop(
```

Comment:
I think this function is not really needed and makes the code harder to grasp. Currently, it is only used twice. In one case `cache` is `None`, in the other `cache` is `Some(_)` and all it does is doing `awaitReady` prior to what it really does, so suggest to use each case in place.

Reply:
Removed, you are right this is better!